fix(robot-server): memory cache analyses

sfoster1 opened the pull request:
Getting analysis results can really bog down the robot server - in fact, it can block the event loop such that new http requests get stalled, and that means that apps will mark the server unhealthy.

Recently, the app started asking for analyses much more frequently, which would cause the robot server to bog down more frequently, which would cause the app to mark the robot unhealthy more frequently. This was an unintended change, and reverting it fixes the robot unhealthiness; but getting analysis results still threatens to bog down the server, so this memcache should fix the problem on the server.

We have similar problems (and solutions) for other heavy resources like runs and protocols; but there we do the caching with functools.lru_cache. This is not possible for analysis results because we get analysis results with async functions, for a very good reason; so instead, we need a custom cache implementation.

This cache implementation uses a python dict to hold the result objects; since there are limited places to add results, we can add things to the cache when they're added to the database and add them to the cache when we get things from the database. We use a deque to maintain a FIFO for cache eviction.

## Review requests
Caches like this aren't really my wheelhouse; think about the behavior here and whether it's correct.

## Testing
- [ ] put this on a robot with a lot of protocols stored and retrieve analyses and see if it falls over

Supercedes #12640

sfoster1 committed:
fix(robot-server): add a cache for analyses

Getting analysis results can really bog down the robot server - in fact,
it can block the event loop such that new http requests get stalled, and
that means that apps will mark the server unhealthy.

Recently, the app started asking for analyses much more frequently, which
would cause the robot server to bog down more frequently, which would
cause the app to mark the robot unhealthy more frequently. This was an
unintended change, and reverting it fixes the robot unhealthiness; but
getting analysis results still threatens to bog down the server, so this
memcache should fix the problem on the server.

We have similar problems (and solutions) for other heavy resources like
runs and protocols; but there we do the caching with
functools.lru_cache. This is not possible for analysis results because
we get analysis results with async functions, for a very good reason; so
instead, we need a custom cache implementation.

This cache implementation uses a python dict to hold the result objects;
since there are limited places to add results, we can add things to the
cache when they're added to the database and add them to the cache when
we get things from the database. We use a deque to maintain a FIFO for
cache eviction.

jbleon95 commented on robot-server/robot_server/protocols/analysis_store.py:
```diff
@@ -386,14 +415,41 @@ async def get_by_protocol(
+        }
+        uncached_analyses = analysis_set - cached_analyses
+
+        if uncached_analyses:
```
Is there a significant performance hit doing this SQL and caching here now? Realize this will only be getting hit once per protocol per session (if I'm reading and understanding this right), but curious if this will be noticeable

sfoster1 replied:
there shouldn't be, I'd think, because this is only getting the ids and not any kind of significant data. The sql result should only be the ids.

jbleon95 commented on robot-server/robot_server/protocols/analysis_store.py:
```diff
@@ -420,6 +476,28 @@ async def add(
+                _log.error(
+                    f"cache order queue was empty with {len(self._completed_analysis_cache_by_analysis_id)} elements in cache"
+                )
+                eldest = "<incorrect because cache order was empty>"
```
Whats the rationale for having this set here and then immediately trying to delete that key from the dictionary, if its just going to raise another error again?

sfoster1 replied:
It makes the logic less complex. Each error check tests precisely one thing. One error case does lead to the other; but in that case, it is truly in that error case. If we hate it, I can factor it into a function that returns early.

SyntaxColoring commented on robot-server/robot_server/protocols/analysis_store.py:
```diff
@@ -386,14 +415,41 @@ async def get_by_protocol(
+                    await _CompletedAnalysisResource.from_sql_row(r)
+                )
+
+        return [
+            self._completed_analysis_cache_by_analysis_id[analysis_id]
+            for analysis_id in ordered_analyses_for_protocol
+        ]
```
This makes me nervous because it depends on none of this protocol's analyses having been evicted from the FIFO by the time we get to this line.

For example: `_CACHE_MAX_SIZE` is 32 right now. If a protocol had 33 analyses, the 1st one would be evicted from `self._completed_analysis_cache_by_analysis_id` by the time we get here, and this would raise a `KeyError`. (This won't happen in practice today because uploading a new protocol is the only thing that triggers an analysis, so the number of analyses per protocol is de-facto capped at 1.)

Or, maybe more realistically: If a bunch of other `get_by_protocol` tasks interrupt us during the `await` on line 446, they'll bump our values out of the FIFO by the time we get here.

sfoster1 replied:
Oh that is a very good point. Might need either a lock or a temporary cache just for this method... I like the latter more

SyntaxColoring commented on robot-server/tests/protocols/test_analysis_store.py:
```diff
@@ -253,3 +256,115 @@ async def test_update_infers_status_from_errors(
     assert analysis.result == expected_result
+
+
+async def test_completed_analyses_use_cache(
```
So far, we've been treating the underlying SQL tables as private implementation details of the `AnalysisStore` et. al. classes.

I think my answer here would be: Don't bother with a pytest test for this. Treat the existence of a cache as an implementation detail.

I think `@mcous`'s answer would be something like:
* Make a "collaborator" class that owns both a "real" store and an in-memory cache, and is responsible for coordinating between them.
* Test the collaborator by instantiating it with a mock real store and a mock cache, and `decoy.verify()`ing that it calls the right methods on them at the right times.
* The existing pytest functions in this file become the unit tests for the real store.
* Write new unit tests for the new cache class.

What do you think?

sfoster1 replied:
Sure, we could do that. that would involve promoting the completed analysis store to a public interface of some other module I guess.

sfoster1 committed:
refactor: refactor for testability

By splitting the implementation to add
- a public module for CompletedAnalysisStore, which has a public
interface that can be tested, with a
- public module for the memory cache, which has a public interface that
can be tested and injected into CompletedAnalysisStore

we can get back to testing only public interfaces, which is really nice.

At the same time, we can fix the race condition where get_by_protocol_id
wasn't as reentrant as it could be. Aside from that method, since this
is async code we should be access safe because the memory cache is synchronous.

SyntaxColoring commented on robot-server/robot_server/protocols/completed_analysis_store.py:
```diff
@@ -0,0 +1,232 @@
+                result = transaction.execute(statement).one()
+            except sqlalchemy.exc.NoResultFound:
+                return None
+        resource = await CompletedAnalysisResource.from_sql_row(
+            result, self._current_analyzer_version
+        )
+        self._memcache.insert(resource.id, resource)
```
When the app was sending "too many" requests to the robot, do you know if there were frequently multiple requests for *the same analysis resource* at a *single instant?*

As written, this can happen:

1. A client asks for a resource
2. The robot starts extracting and parsing the resource
3. A client (maybe the same one, maybe different) asks for the same resource again
4. **The robot starts extracting and parsing the resource a second time, redundantly, while the first extract+parse is still ongoing**
5. The robot returns the resource to first client, caching it
6. The robot returns resource to second client

Which is totally good for this PR, but I'm curious how much we need to anticipate an optimization like this in the future:

1. A client asks for a resource
2. The robot starts extracting and parsing the resource
3. A client (maybe the same one, maybe different) asks for the same resource again
4. **The robot knows it's already extracting and parsing that resource, so it waits for that task to complete instead of starting a new one.**
5. The robot returns the resource to first client, caching it
6. The robot returns resource to second client

sfoster1 replied:
That is definitely something we could do - though regardless of the timing and status of this pr, I think it's functionality separate enough that it should be in a different one - but I think it would be some tough sledding to show correctness. I think the stages we'd want to go through for this are:
1. Make the cache bigger
2. See if holding a route-scope lock I guess keyed on the protocol id after pulling analysis ids and before checking cache status is good enough; this is the right place because that's when we have the information we'd need, and we would need the cache examination to stay async-atomic to retrieving things so that waiters could see the cache state _after_ the related sql commands
3. Bounce sql tasks into a by-analysis-id store or something and have access to that locked

I think we get analysis rarely enough that 2 would be enough.

In terms of how often that happens, I think the answer is pretty often _during development_, where there's a lot of the flow of like 
- some people are testing something on a robot
- they make some change 
- the robot server boots
- they both click the devices page
(ignoring how the app changed to not do this anymore).  I think that flow is a little less likely in general use.

SyntaxColoring commented on robot-server/robot_server/protocols/completed_analysis_store.py:
```diff
@@ -0,0 +1,232 @@
+        # for this coroutine - the cache class itself is light, python is pass-by-reference,
+        # so we shouldn't incur much extra memory, and since we're using an unlimited cache
+        # that bypasses the size checks we should be pretty cheap timewise also.
+        local_memcache = MemoryCache(
+            max(len(ordered_analyses_for_protocol), 1), str, CompletedAnalysisResource
+        )
```
Instead of using a local `MemoryCache` here with a carefully-selected size limit, it seems like it would be simpler to use a plain `dict` here?

sfoster1 replied:
sure

SyntaxColoring commented on robot-server/tests/protocols/test_completed_analysis_store.py:
```diff
@@ -0,0 +1,188 @@
+    # return the identity-same resource
+    decoy.when(memcache.get("analysis-id")).then_return(resource)
+    assert (await subject.get_by_id("analysis-id")) is resource
+    decoy.verify(memcache.get("analysis-id"))
```
One of Decoy's opinions is that we should omit this `.verify()` because this check is already covered by your`decoy.when(...).then_return(resource)`+`assert (...) is resource`. Decoy is raising a warning about this:

```
RedundantVerifyWarning: The same rehearsal was used in both a `when` and a `verify`.
  This is redundant and probably a misuse of the mock.
  	MemoryCache.get('analysis-id')
  See https://mike.cousins.io/decoy/usage/errors-and-warnings/#redundantverifywarning
    warn(RedundantVerifyWarning(rehearsal=vr))
```

In this case, I think we can fix it without taking any power away from the test by just removing the `.verify()`.

```suggestion
```

SyntaxColoring commented on robot-server/robot_server/protocols/analysis_memcache.py:
```diff
@@ -0,0 +1,65 @@
+    _cache_order: Deque[K]
+    _cache_size: int
+
+    def __init__(self, size_limit: int, _keyhint: Type[K], _valhint: Type[V]) -> None:
```
Just for my edification, no opinion from me:

If this constructor didn't have `_keyhint` and `_valhint`, you could construct `MemoryCache` like `MemoryCache[str, Foo](123)`. Is there a reason you're preferring the `MemoryCache(123, str, Foo)` syntax instead?

sfoster1 replied:
It's entirely preference based. I hate typing out type hints. If you hate it, I can change it back.

SyntaxColoring replied:
Nope, this works for me.

TamarZanzouri reviewed (approved):
I love this change!!

SyntaxColoring commented on robot-server/tests/protocols/test_completed_analysis_store.py:
```diff
@@ -0,0 +1,188 @@
+    assert analysis_from_sql == resource
+
+
+async def test_get_by_analysis_id_caches_results(
+    subject: CompletedAnalysisStore,
+    sql_engine: Engine,
+    memcache: MemoryCache[str, CompletedAnalysisResource],
+    protocol_store: ProtocolStore,
+    decoy: Decoy,
+) -> None:
+    """It should cache successful fetches from sql."""
+    resource = _completed_analysis_resource("analysis-id", "protocol-id")
+    protocol_store.insert(make_dummy_protocol_resource("protocol-id"))
+    await subject.add(resource)
+    # the analysis is not cached
+    decoy.when(memcache.get("analysis-id")).then_raise(KeyError())
+    from_sql = await subject.get_by_id("analysis-id")
+    assert from_sql == resource
+    decoy.verify(memcache.insert("analysis-id", from_sql))
```
This is a departure from Robot Services' test and architectural conventions, but it's a departure that I've been thinking we should try, anyway. No action needed—I'll dump my thoughts in Slack.

SyntaxColoring reviewed (approved):
Thanks! This looks good to me architecturally. Here are a few low-level suggestions.

SyntaxColoring reviewed (approved):
Thanks! This looks good to me architecturally. Here are a few low-level suggestions.

sfoster1 committed:
Update robot-server/tests/protocols/test_completed_analysis_store.py

Co-authored-by: Max Marrone <max@opentrons.com>

sfoster1 committed:
use a local dict instead of full memcache

SyntaxColoring commented:
Smoke-tested on a robot by navigating around. The `journalctl` logs seem to confirm that the cache is correctly eliding database access.

sfoster1 commented:
> Smoke-tested on a robot by navigating around. The `journalctl` logs seem to confirm that the cache is correctly eliding database access.

Thank you!
